internal/hcsoci/devices.go

```diff
@@ -0,0 +1,108 @@
+package hcsoci
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/Microsoft/hcsshim/internal/devices"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/oci"
+	"github.com/Microsoft/hcsshim/internal/resources"
+	"github.com/Microsoft/hcsshim/internal/uvm"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+)
+
+const deviceUtilExeName = "device-util.exe"
+
+// getAssignedDeviceKernelDrivers gets any device drivers specified on the spec.
+// Drivers are optional, therefore do not return an error if none are on the spec.
+//
+// See comment on oci.AnnotationAssignedDeviceKernelDrivers for expected format.
+func getAssignedDeviceKernelDrivers(annotations map[string]string) ([]string, error) {
+	csDrivers, ok := annotations[oci.AnnotationAssignedDeviceKernelDrivers]
+	if !ok || csDrivers == "" {
+		return nil, nil
+	}
+	drivers := strings.Split(csDrivers, ",")
+	for _, driver := range drivers {
+		if _, err := os.Stat(driver); err != nil {
+			return nil, errors.Wrapf(err, "failed to find path to drivers at %s", driver)
+		}
+	}
+	return drivers, nil
+}
+
+// getDeviceUtilHostPath is a simple helper function to find the host path of the device-util tool
+func getDeviceUtilHostPath() string {
+	return filepath.Join(filepath.Dir(os.Args[0]), deviceUtilExeName)
+}
+
+// handleAssignedDevicesWindows does all of the work to setup the hosting UVM, assign in devices
+// specified on the spec, and install any necessary, specified kernel drivers into the UVM.
+//
+// Drivers must be installed after the target devices are assigned into the UVM.
+// This ordering allows us to guarantee that driver installation on a device in the UVM is completed
+// before we attempt to create a container.
```

Comment:
I think it would be good to add a comment here which says that this function takes in the `WindowsDevice` spec provided in the container config and returns a modifies copy of this spec by including the location paths of the assigned devices and this modified spec should actually be passed to HCS while creating a container with the assigned devices.